### ISSUE TITLE
VPNaaS - api calls

### ISSUE THREAD

**Issue by shuwens** [shwsun/haas#4]
Design rest api call, say `project_make_vpn` to make a Vpn for the project requires it, `project_destroy_vpn` to kill process and destroy config files, etc. 

REST api call:
- [ ] `project_make_vpn`
- [ ] `project_destroy_vpn`

---

**project_make_vpn**
1. get the project/network of the project that requires VPN
2. create br-vlan that attach to the VLAN/network channel on the vpn_node
3. create tap device and put the tap on the br-vlan

**project_destroy_vpn**
*will be some security concern, how to make sure some user won't use used vpn to access others' project.*
1. destroy all the certs?

**vpn_spawn**
*since generating cert/ dh pem will cost some time, it might be that case that we want to do this in advance, s.t. `project_make_vpn` will only in charge of allocate vpn to projects*
1. create cert, user can specify what level of security required, will be 2048 by default
2. create tap dev

**Comment by matsuuran** [comment]
@shwsun do you want me to knock out calls today/tomorrow?  Sorry I've been busy all week and didn't notice this

**Comment by shuwens** [comment]
Yeah if you feel like want to get sth done, just go ahead and do it.

**Comment by zenhack** [comment]
Here's my take: we have two API calls: `create_vpn` and `destroy_vpn`

`create_vpn` takes a network, a project, and static openvpn key. It allocates a UDP port out of a pool, saves the key, generates an openvpn config file, starts the openvpn process for that vpn, and informs the user of what host/port they need to connect to.

Authorization requirements: Access to both the project and the network (or admin access).

`destroy_vpn` stops the openvpn process, deletes the config file and the static key, and returns the port to the pool.

I _believe_ openvpn can be configured to manage the tap device itself, but if not we can create a bunch on startup and allocate from a pool.

Some notes:
- Having the client supply the key has the advantage that we don't have to worry about key generation consuming server resources, or causing latency issues.
- Using static keys means that we can just delete them to revoke access, instead of having to mess with PKI stuff.
- Using a dedicated daemon (e.g. systemd) to manage the openvpn processes' lifecycle will make it easy to be reasonably confident that the networks have actually been shut down properly.

**Comment by shuwens** [comment]
@zenhack I agree on the point that let clients create the certs, and it will help us avoid PKI issue. What I am concerned is that openvpn requires 4 files(ca cert, cert, key, dh) to start a vpn service. Do you think it is right approach to have 6 params in a api call? o/w we will do some parsing in the api call.  (e.g parse one file with those 4 fields)

Also, we are doing br-vlan to serve headnodes talking to nodes on the network channel, I assume we need to do the same thing for the VPNnode, right? This part and the virtual nic is the mechanisms that I said we might keep.

**Comment by zenhack** [comment]
Quoting (Jethro) Shuwen Sun (2016-11-16 11:56:35)

>    @zenhack I agree on the point that let clients create the certs, and
>    it will help us avoid PKI issue. What I am concerned is that openvpn
>    requires 4 files(ca cert, cert, key, dh) to start a vpn service. Do you
>    think it is right approach to have 6 params in a api call? o/w we will
>    do some parsing in the api call. (e.g parse one file with those 4
>    fields)

The ca cert doesn't exist for a static key -- certificate authorities
are a PKI thing. I think we should just need one file, see:

https://openvpn.net/index.php/open-source/documentation/miscellaneous/78-static-key-mini-howto.html

>    Also, we are doing br-vlan to serve headnodes talking to nodes on the
>    network channel, I assume we need to do the same thing for the VPNnode,
>    right? This part and the virtual nic is the mechanisms that I said we
>    might keep.

The bridges and the VLAN nics would stay as-is, yes.

**Comment by matsuuran** [comment]
@zenhack @shwsun  do you guys have any thoughts on how we should pass the key file?

I've been looking at libraries that can send files (pycurl and pysendfile namely) but wanted to get input before moving forward

maybe since I have to use subprocesses to start openvpn (right?) it may be best to just use a subprocess with `scp`

**Comment by zenhack** [comment]
Pass it where? It will be embedded in the http request for the api call, and then the server writes it to a file, and puts the filename in openvpn's config.

**Comment by matsuuran** [comment]
Is the API server always going to be the same as the server that has the VPN nodes?  I am under the impression that we would want the key to reside on the VPN node

**Comment by zenhack** [comment]
That had been my thinking, at least for the time being. We may want to split it out into a separate service at some point, but for now I wouldn't add the extra complexity. @henn, interested in your thoughts on this.

I do have some more general thoughts about splitting off bits of HIL functionality. I want to do the console stuff as a separate service, in part because I think it's going to be a pain to integrate into the API server. but, one thing at a time.

**Comment by matsuuran** [comment]
For the cli call `project_vpn_create` I think it would be better to pass the path to the key and then read the key that way to put it in the http request.  This way the user doesn't have a really long command.

This sort of begs the questions, because the cli and api are running on the same server wouldn't it be even easier to just pass the key path to the api and then move the file to the correct location with the api?

**Comment by zenhack** [comment]
The CLI is *not* necessarily running on the same machine.

Even if it were, it likely wouldn't be running as the same user, which means it wouldn't necessarily be able to access the named file, and furthermore could easily result in confused deputy attacks where the API server ends up reading some arbitrary file that the CLI isn't supposed to have access to. 

I think having the CLI take in a file path and read the key from the file is a good idea.

**Comment by shuwens** [comment]
> I do have some more general thoughts about splitting off bits of HIL functionality. I want to do the console stuff as a separate service, in part because I think it's going to be a pain to integrate into the API server. but, one thing at a time.

Regarding this part, during the last HIL weekly meeting we kinda agree to put VPN stuff in `ext/`. Such that we can hook it into our deployment test env(any changes that break it will break HIL). Your thoughts?

**Comment by zenhack** [comment]
Yeah, I think putting it in `ext/` is the way to go for now.

**Comment by matsuuran** [comment]
@shwsun do you have the server.conf file we're going to use?

**Comment by shuwens** [comment]
@matsuuran  just as what Ian has pointed out, the simple config file should work 
https://openvpn.net/index.php/open-source/documentation/miscellaneous/78-static-key-mini-howto.html

**Comment by matsuuran** [comment]
I'm running into a problem with the vpn@.service script.  Essentially to use it I'd have to put the key and the script into `/etc/openvpn/`.  I think we're going to run into permissions issues unless there is something I'm missing.

I have not found an elegant way to copy files to a folder owned by root, but if you know of one please let me know.

**Comment by zenhack** [comment]
We may need to write our own service script, if only to have the files stored elsewhere. And this is going to be running in a systemd instance owned by the HIL user, not the system-wide one (since we don't want access to that if we can avoid it). Also, if this isn't already on folks' radar: we'll probably have to pre-create the tap devices, since that's a priviledged operation as well.
